File: lib/djs.js
var WS;
var REFS;
var SERVER;
var CALLBACKS;
var handshaked;
var isWebsocket;
var TASKS;
var EVENTS;
var FUNCS;

// WebSocket Implementation
function djs_start_websocket(server) {
    WS = new WebSocket(server);
    WS.onopen = on_open;
    WS.onclose = on_close;
    WS.onmessage = on_message;
    WS.onerror = on_error;
    handshaked = false;
    REFS = new Array();
    TASKS = new Array();
    CALLBACKS = new Array();
    isWebsocket = true;
    EVENTS = new Array();
    FUNCS = new Array();
}

function websocket_handshake() {
    WS.send("0\t0\t");
}

function on_open() {
    websocket_handshake();
    console.log("opened");
}

function on_close() {
    console.log("closed");
}

function on_message(event) {
    if (!handshaked) {
        handshaked = true;
        DJS_ID = event.data;
        WS.send("1\t" + DJS_ID + "\t");
    } else {
        if (event.data == "\t") {
            return;
        }
        var rsp = djs_eval(event.data);
        WS.send("2\t" + DJS_ID + "\t" + rsp);
    }
}

function on_error(error) {
    console.log("error");
}

function oncallback(cid, method, event) {
    WS.send("5\t" + cid + "\t" + method);
}

// AJAX Implementation
function djs_start_ajax(server) {
    SERVER = server;
    REFS = new Array();
    TASKS = new Array();
    CALLBACKS = new Array();
    isWebsocket = false;
    EVENTS = new Array();
    FUNCS = new Array();

    handshake();
}

function djs_send(msg, callback) {
    message(msg, callback);
}

function getXmlHttpRequestObject() {
    var xhr;
    if (XMLHttpRequest) {
        xhr = new XMLHttpRequest();
    } else {
        try {
            xhr = new ActiveXObject('MSXML2.XMLHTTP.6.0');
        } catch (e) {
            try {
                xhr = new ActiveXObject('MSXML2.XMLHTTP.3.0');
            } catch (e) {
                try {
                    xhr = new ActiveXObject('MSXML2.XMLHTTP');
                } catch (e) {
                    xhr = null;
                    alert("This browser does not support XMLHttpRequest.");
                }
            }
        }
    }

    return xhr;
}

function handshake() {
    var xhr = getXmlHttpRequestObject();
    xhr.open("POST", SERVER);
    xhr.setRequestHeader("Content-Type", "text/plain");
    xhr.onreadystatechange = function() {
        if (xhr.readyState === 4 && xhr.status === 200) {
            if (xhr.responseText == "\t") {
                return;
            }
            connect(xhr.responseText);
        }
    };
    xhr.send("0\t0\t");
}

function connect(id) {
    var xhr = getXmlHttpRequestObject();
    xhr.open("POST", SERVER);
    xhr.setRequestHeader("Content-Type", "text/plain");
    xhr.onreadystatechange = function() {
        if (xhr.readyState === 4 && xhr.status === 200) {
            var rsp = xhr.responseText;
            if (rsp == "\t") {
                return;
            }
            if (rsp == "") {
                reconnect(id);
                return;
            }
            var req = djs_eval(rsp);
            response(id, req);
        }
    };
    xhr.send("1\t" + id + "\t");
}

function reconnect(id) {
    var xhr = getXmlHttpRequestObject();
    xhr.open("POST", SERVER);
    xhr.setRequestHeader("Content-Type", "text/plain");
    xhr.onreadystatechange = function() {
        if (xhr.readyState === 4 && xhr.status === 200) {
            var rsp = xhr.responseText;
            if (rsp == "\t") {
                return;
            }
            if (rsp == "") {
                reconnect(id);
                return;
            }
            var req = djs_eval(rsp);
            response(id, req);
        }
    };
    xhr.send("6\t" + id + "\t");
}

function response(id, rsp) {
    var xhr = getXmlHttpRequestObject();
    xhr.open("POST", SERVER);
    xhr.setRequestHeader("Content-Type", "text/plain");
    xhr.onreadystatechange = function() {
        if (xhr.readyState === 4 && xhr.status === 200) {
            var rsp = xhr.responseText;
            if (rsp == "\t") {
                return;
            }
            if (rsp == "") {
                reconnect(id);
                return;
            }
            var req = djs_eval(rsp);
            response(id, req);
        }
    };
    xhr.send("2\t" + id + "\t" + rsp);
}

function callback(cid, method, event) {
    var xhr = getXmlHttpRequestObject();
    xhr.open("POST", SERVER);
    xhr.setRequestHeader("Content-Type", "text/plain");
    xhr.onreadystatechange = function() {
        if (xhr.readyState === 4 && xhr.status === 200) {
            var rsp = xhr.responseText;
            var id = rsp.substring(0, rsp.indexOf("\t"));
            if (rsp == "\t") {
                return;
            }
            var req = djs_eval(rsp.substring(rsp.indexOf("\t") + 1));
            response(id, req);
        }
    }
    var event_id = EVENTS.length;
    EVENTS[event_id] = event;
    xhr.send("5\t" + cid + "\t" + method + "\t" + event_id);
}

function rpc(id) {
    var xhr = getXmlHttpRequestObject();
    xhr.open("POST", SERVER);
    xhr.setRequestHeader("Content-Type", "text/plain");
    xhr.onreadystatechange = function() {
        if (xhr.readyState === 4 && xhr.status === 200) {
            var rsp = xhr.responseText;
            if (rsp == "\t") {
                return;
            }
            var req = djs_eval(rsp);
            response(id, req);
        }
    }
    xhr.send("7\t" + id + "\t");
}

function djs_eval(json, ext) {
    var rsp = "";
    try {
        var reqs = json.split("\t");
        var info;
        var id;
        var cid;
        var obj;
        var content;
        var args;
        var result;
        var i;
        var ex;
        if (reqs[0] == "1") {
            for (i = 1; i < reqs.length; i++) {
                info = eval("(" + reqs[i] + ")");
                REFS[info.id].origin = info.origin;
            }
        } else if (reqs[0] == "2") {
            var func = json.substring(2, json.length);
            var nn = func.split(",", 1)[0];
            var cc = func.substring(nn.length + 1);
            createFunction(nn, cc);
            return;
        } else {
            for (i = 1; i < reqs.length; i++) {
                TASKS.push(reqs[i]);
            }
        }

        var req;
        rsp = "{";
        while (TASKS.length > 0) {
            req = TASKS.shift();
            info = eval("(" + req + ")");
            id = info.id;
            cid = info.cid;
            obj = info.type;
            content = info.content;
            args = info.args;
            ex = info.ex;

            if (obj == "rpc") {
                rpc(content);
                continue;
            }

            if (obj == "new") {
                var instance = "new " + content + "(";
                if (args && args.length > 0) {
                    instance += "args[0]";
                    for (i = 1; i < args.length; i++) {
                        instance += ",args[" + i + "]";
                    }
                }
                instance += ")";
                createRefObj(id, eval(instance));
                continue;
            }

            if (content.indexOf("{}") == 0) {
                obj[content.substring(2)] = createCallback(cid, info.args[0]);
                result = null;
            } else if (content == "[]") {
                // TODO args
                obj = to_ruby_object(id, obj[args[0]]);
                if (obj) {
                    rsp += id + "=>" + obj + ",";
                }
            } else if (content == "[]=") {
                obj[args[0]] = args[1];
            } else {
                if (!ext && isPrimitive(obj) && !obj[content]) {
                    to_ruby_object(info.id, null);
                    info.type = obj;
                    return toRubyHash(info);
                }

                if (args && args.length > 0) {
                    if (content.match(/^.+=$/)) {
                        var prop = content.substring(0, content.length - 1);
                        obj[prop] = args[0];
                        result = null;
                    } else {
                        if (!obj[content]) {
                            throw new Error("undefined method: '" + content + "'");
                        }
                        result = obj[content].apply(obj, args);
                    }
                } else {
                    result = obj[content];
                    if (result == undefined) {
                        throw new Error("undefine method or property: '" + content + "'");
                    }
                    if (result && typeof result == "function" && !(ex == "object")) {
                        result = result.call(obj);
                    }
                }
            }

            if (ext) {
                createRefObj(id, result);
            } else {
                obj = to_ruby_object(id, result);
                if (obj) {
                    rsp += id + "=>" + obj + ",";
                }
            }
        }

        if (rsp.length > 1) {
            rsp = rsp.substring(0, rsp.length - 1) + "}";
        } else {
            rsp += "}";
        }
    } catch (e) {
        rsp = "{'error'=>" + to_ruby_object(-1, e.message) + ",'id'=>" + id + "}";
    }
    return rsp;
}

function createRefObj(id, obj) {
    var refObj = new RefObj(id, obj);
    REFS[id] = refObj;
    return refObj;
}

function createCallback(cid, method) {
    return function(cid, method) {
        return function(e) {
            if (FUNCS[method]) {
                FUNCS[method](e);
            } else {
                if (isWebsocket) {
                    oncallback(cid, method, e);
                } else {
                    callback(cid, method, e);
                }
            }
        }
    }(cid, method);
}

function createFunction(name, content) {
    FUNCS[name] = function() {
        djs_eval(content, arguments);
    }
}

function RefObj(id, obj) {
    this.id = id;
    this.origin = obj;
}

function to_ruby_object(id, obj) {
    if (obj == undefined || obj == null) {
        return null;
    }

    var type = typeof obj;
    if (type == "string") {
        return "'" + obj.replace(/'/g, "\\'") + "'";
    } else if (type == "number" || type == "boolean") {
        return obj;
    } else {
        createRefObj(id, obj);
        return null;
    }
}

function isPrimitive(obj) {
    var type = typeof obj;
    return type == "string" || type == "number" || type == "boolean";
}

function toRubyHash(hash) {
    var ret = '{';
    for (key in hash) {
        ret += '"' + key + '"=>';
        if (typeof hash[key] == "string") {
            ret += '"' + hash[key] + '",';
        } else {
            if (hash[key] instanceof Array && hash[key].length == 0) {
                ret += '[],';
            } else {
                ret += hash[key] + ',';
            }
        }
    }
    if (ret.length > 1) {
        ret = ret.substring(0, ret.length - 1);
    }
    ret += '}';
    return ret;
}

File: lib/djs.rb
require 'thread'

module DJS
  TYPE_HANDSHAKE = 0
  TYPE_CONNECT   = 1
  TYPE_RESPONSE  = 2
  TYPE_MESSAGE   = 3
  TYPE_CLOSE     = 4
  TYPE_CALLBACK  = 5
  TYPE_RECONNECT = 6
  TYPE_RPC       = 7

  SPECIAL_METHODS = [:childNodes]

  TERMINAL = "\t"

  class DJSError < RuntimeError; end
  class DJSBadRequestError < DJSError; end

  def self.start(klass = DJSConnection, opts = {})
    @@server = DJSServer.new(klass, opts)
    DJSConnections.server = @@server
  end

  def self.stop
    @@server.stop
  end

  def self.response(request)
    if request =~ /\A(\d)\t(\d+)\t(.*)\z/
      case $1.to_i
      when TYPE_HANDSHAKE
        response = @@server.handshake
      when TYPE_CONNECT
        response = @@server.connect($2.to_i)
      when TYPE_RESPONSE
        response = @@server.response($2.to_i, $3)
      when TYPE_MESSAGE
        # TODO
      when TYPE_CLOSE
        # TODO
      when TYPE_CALLBACK
        response = @@server.callback($2.to_i, *$3.split("\t"))
      when TYPE_RECONNECT
        response = @@server.reconnect($2.to_i)
      when TYPE_RPC
        response = @@server.rpc($2.to_i)
      else

      end
    else
      Thread.new do
        onerror DJSBadRequestError.new
      end
      response = TERMINAL
    end

    return response.to_s
  end

  def self.connections
    @@server.djs_connections
  end

  def self.add_task(conn, name, *args)
    @@server.add_task(conn, name, *args)
  end

  class DJSConnections

    def self.server=(server)
      @@server = server
    end

    def initialize(connections)
      @connections = connections
    end

    def [](pattern)
      if pattern == "*"
        return self
      end
      DJSConnections.new(Hash.new)
    end

    def method_missing(name, *args)
      @connections.values.each { |conn|
        next unless conn.type == :main
        rpc_conn = @@server.create_rpc_connection(conn.cid, name, args)
        prx = DJSRpcProxyObject.new(rpc_conn)
        conn.request.push "0\t" + prx.__json
      }
    end
  end

  class DJSServer
    RECONNECT = ""
    DEFAULT_OPTS = {
      :buff_size => 200
    }

    def initialize(klass = DJSConnection, opts = {})
      @klass = klass
      @opts = DEFAULT_OPTS.merge(opts)

      @connections = Hash.new
      @djs_connections = DJSConnections.new(@connections)
      @tasks = Queue.new
      @group = ThreadGroup.new
      @thread = run
      @group.add @thread
    end
    attr_reader :connections, :djs_connections

    def run
      Thread.start do
        begin
          while true
            main_loop
          end
        ensure
          # TODO kill thread
        end
      end
    end

    def main_loop
      Thread.start(@tasks.pop) do |task|
        conn = task[0]
        name = task[1]
        args = task[2]
        fiber = Fiber.new do
          if args
            conn.send(name, *args)
          else
            conn.send(name)
          end
          conn.add_proxy(nil, true)
          TERMINAL
        end
        req = fiber.resume
        if req == TERMINAL
          if conn.type == :main
            conn.request.push RECONNECT
          else
            @connections.delete(conn.__id__)
            conn.request.push TERMINAL
            return
          end
        else
          conn.request.push req
        end
        while rsp = conn.response.pop
          req = fiber.resume(rsp)
          if req == TERMINAL
            if conn.type == :main
              conn.request.push RECONNECT
            else
              @connections.delete(conn.__id__)
              conn.request.push TERMINAL
              return
            end
          else
            conn.request.push req
          end
        end
      end
    end

    def stop
      @group.list.each { |thread|
        thread.kill
      }
    end

    def handshake
      @group.add Thread.current
      conn = @klass.new(:main, 0, @opts)
      @connections[conn.__id__] = conn
      return conn.__id__.to_s
    end

    def connect(id)
      @group.add Thread.current
      conn = @connections[id]
      add_task conn, :on_open
      req = conn.request.pop
      return req
    end

    def reconnect(id)
      @group.add Thread.current
      conn = @connections[id]
      req = conn.request.pop
      return req
    end

    def response(id, rsp)
      @group.add Thread.current
      conn = @connections[id]
      conn.response.push rsp
      req = conn.request.pop
      return req
    end

    def callback(cid, method, event_id)
      @group.add Thread.current
      conn = @klass.new(:callback, cid, @opts)
      @connections[conn.__id__] = conn
      add_task conn, :on_callback, method, event_id
      req = conn.request.pop
      return conn.__id__.to_s + "\t" + req
    end

    def rpc(id)
      @group.add Thread.current
      conn = @connections[id]
      add_task conn, conn.rpc[0], *conn.rpc[1]
      req = conn.request.pop
      return req
    end

    def add_task(conn, name, *args)
      @tasks.push [conn, name, args]
    end

    def create_rpc_connection(cid, name, args)
      conn = @klass.new(:rpc, cid, @opts)
      @connections[conn.__id__] = conn
      conn.set_rpc(name, args)
      return conn
    end
  end

  class DJSConnection
    SEP = "\t"

    def initialize(type, cid = 0, opts = {})
      @type = type
      if @type == :main
        @cid = self.__id__
      else
        @cid = cid
      end
      @opts = opts
      @request = Queue.new
      @response = Queue.new
      @proxies = Hash.new
      @proxy_ids = Array.new
      @mutex = Mutex.new
      @registered_function = Array.new
    end

    attr_accessor :type, :request, :response, :proxies
    attr_reader :cid, :registered_function

    def set_rpc(name, args)
      @rpc = [name, args]
    end

    def rpc
      @rpc
    end

    def add_proxy(proxy, flush_now = false)
      if @register
        @register << SEP << proxy.__json if proxy
      else
        @mutex.synchronize {
          if proxy
            @proxy_ids << proxy.__id__
            @proxies[proxy.__id__] = proxy
          end
          # TODO length
          if flush_now || @proxy_ids.length == @opts[:buff_size]
            flush
          end
        }
      end
    end

    def sync
      add_proxy(nil, true)
    end

    def flush
      if @proxy_ids.empty?
        return
      end

      error = nil

      json = "0\t"
      @proxy_ids.each { |id|
        json << @proxies[id].__json << SEP
      }
      json[-1] = ""
      rsp = Fiber.yield(json)
      while rsp != '{}'
        info = eval(rsp)

        if info.key?('type')
          result = info['type'].send(info['content'], *info['args'])
          proxy = @proxies[info['id']]
          proxy = @proxies[proxy.info[:id]]
          proxy.origin = result
          proxy.solved = true

          json = "1\t{\"id\":" + DJS.to_json(info["id"]) + ",\"origin\":" + DJS.to_json(result) + "}"
          rsp = Fiber.yield(json)
        elsif info.key?('error')
          error = @proxies[info['id']].exe_info + ": " + info['error']
          break
        else
          info.each { |key, value|
            if @proxies.key?(key)
              @proxies[key].origin = value
              @proxies[key].solved = true
            end
          }
          break
        end
      end

      @proxies.clear
      @proxy_ids.clear

      if error
        on_error(error)
      end
    end

    def on_callback(method_name, event_id)
      method = self.method(method_name)
      if method.arity == 1
        event = DJSEventProxyObject.new(self, event_id)
        method.call(event)
      else
        method.call
      end
    end

    # Event handler
    def on_open

    end

    def on_close

    end

    def on_message(msg)

    end

    def on_error(err)

    end

    # Root Javascript Object
    def window
      DJSProxyObject.new(self, {:type => :window})
    end

    def register_function(name)
      sync
      method = self.method(name)
      @register = '2' << SEP << name.to_s << ",0"
      if method.arity < 0
        raise "Optional arguments is not supported!"
      end
      arguments = []
      method.arity.times { |i|
        arguments << DJSArgument.new(self, "ext[#{i}]")
      }
      method.call(*arguments)
      Fiber.yield(@register)
      @register = nil
      @registered_function << name
    end

    def method_missing(name, *args, &block)
      window.method_missing(name, *args)
    end
  end

  class DJSProxyObject
    def initialize(conn, info, exe_info = '')
      @conn = conn
      @info = info
      @origin = nil
      @solved = false
      @info[:id] = __id__
      @info[:cid] = conn.cid
      @exe_info = exe_info
    end

    attr_accessor :origin, :solved, :info, :conn
    attr_reader :exe_info

    def method_missing(name, *args, &block)
      if @solved
        return @origin.send(name, *args, &block)
      end

      if block
        @conn.add_proxy(nil, true)
        return @origin.send(name, *args, &block)
      end

#      if @conn.registered_function.include?(name)
#        proxy = DJSProxyObject.new(@conn, {:type => "window", :content => "FUNCS[#{name}]", :args => args}, caller[0])
#        @conn.add_proxy proxy
#        return proxy
#      end

      if name.to_s =~ /^_[A-Z]/ && args.size == 0
        return DJSClassProxyObject.new(@conn, name, caller[0])
      end

      if name.to_s =~ /^.*=$/ && args && Symbol === args[0]
        name = "{}" + name.to_s[/[^=]+/]
      end

      if SPECIAL_METHODS.include?(name)
        proxy = DJSProxyObject.new(@conn, {:type => self, :content => name, :args => args, :ex => 'object'}, caller[0])
      else
        proxy = DJSProxyObject.new(@conn, {:type => self, :content => name, :args => args}, caller[0])
      end
      @conn.add_proxy proxy
      return proxy
    end

    def [](key)
      proxy = DJSProxyObject.new(@conn, {:type => self, :content => :[], :args => [key]}, caller[0])
      @conn.add_proxy proxy
      return proxy
    end

    def []=(key, value)
      proxy = DJSProxyObject.new(@conn, {:type => self, :content => :[]=, :args => [key, value]}, caller[0])
      @conn.add_proxy proxy
      return proxy
    end

    def sync
      return @origin if @solved
      @conn.add_proxy(nil, true)
      return @solved ? @origin : self
    end

    def __json
      DJS.to_json @info
    end

    def __to_s
      #TODO refs
      if DJSProxyObject === @info[:type]
        return "REFS[" + __id__.to_s + "].origin"
      else
        return @info[:type].to_s
      end
    end
  end

  class DJSEventProxyObject < DJSProxyObject
    def initialize(conn, event_id)
      super(conn, {:type => "EVENTS[#{event_id}]"})
    end
  end

  class DJSRpcProxyObject < DJSProxyObject
    def initialize(conn)
      super(conn, {:type => 'rpc', :content => conn.__id__})
    end
  end

  class DJSArgument < DJSProxyObject
    def initialize(conn, name)
      @name = name
      super(conn, {:type => self})
    end

    def __to_s
      @name
    end
  end

  class DJSClassProxyObject < DJSProxyObject
    def initialize(conn, name, caller)
      @name = name[1, name.length]
      super(conn, {:type => self}, caller)
    end

    def new(*args)
      proxy = DJSProxyObject.new(@conn, {:type => 'new', :content => @name, :args => args}, caller[0])
      @conn.add_proxy proxy
      proxy.info[:type] = self
      return proxy
    end
  end

  def self.to_json(obj)
    case obj
    when Hash
      json = '{'
      obj.each { |key, value|
        json << '"' << key.to_s << '":' << to_json(value) << ','
      }
      if (json.length > 1)
        json[-1] = '}'
      else
        json << '}'
      end
      return json
    when Array
      json = '['
      obj.each { |item|
        json << to_json(item) << ','
      }
      if (json.length > 1)
        json[-1] = ']'
      else
        json << ']'
      end
      return json
    when DJSProxyObject
      return obj.__to_s
    when Numeric
      return obj.to_s
    when String
      return '"' + obj.gsub('"', '\\"') + '"'
    when TrueClass, FalseClass
      return obj.to_s
    else
      return '"' + obj.to_s + '"'
    end
  end
end
